ethereum/bridge/contracts/Bridge.sol

```diff
@@ -0,0 +1,295 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.20;
+
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import '@openzeppelin/contracts/utils/cryptography/ECDSA.sol';
+import '@openzeppelin/contracts/utils/ReentrancyGuard.sol';
+import '@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol';
+import '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import '@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol';
+import {ERC721Upgradeable} from '@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol';
+
+import './ChainIDs.sol';
+import './TokenIDs.sol';
+
+// Bridge contract
+contract Bridge is Initializable, UUPSUpgradeable, ERC721Upgradeable, ChainIDs {
+	using SafeERC20 for IERC20;
+	using MessageHashUtils for bytes32;
+
+	uint256[48] __gap;
+
+	mapping(address => mapping(uint => bool)) public processedNonces;
+
+	// uint8 private immutable version;
+	// uint8 private version;
+
+	bool public paused;
+
+	uint16 public constant MAX_TOTAL_WEIGHT = 10000;
+	uint256 public constant MAX_SINGLE_VALIDATOR_WEIGHT = 1000;
+	uint256 public constant APPROVAL_THRESHOLD = 3333;
+
+	// A struct to represent a validator
+	struct Validator {
+		address addr; // The address of the validator
+		uint256 weight; // The weight of the validator
+	}
+
+	struct ApprovedBridgeMessage {
+		BridgeMessage message;
+		uint64 approvedEpoch;
+		bytes[] signatures;
+	}
+
+	struct BridgeMessageKey {
+		uint8 sourceChain;
+		uint64 bridgeSeqNum;
+	}
+
+	struct BridgeMessage {
+		// 0: token , 1: object ? TBD
+		uint8 messageType;
+		uint8 version;
+		ChainID sourceChain;
+		uint64 bridgeSeqNum;
+		address senderAddress;
+		uint8 targetChain;
+		address targetAddress;
+		// bytes payload;
+	}
+
+	// A mapping from address to validator index
+	mapping(address => uint256) public validatorIndex;
+
+	// An array to store the validators
+	Validator[] public validators;
+
+	// Mapping of user address to nonce
+	mapping(address => uint256) public nonces;
+
+	// Event to emit when a transfer is initiated
+	event TransferInitiated(
+		address indexed sender,
+		address indexed recipient,
+		uint256 amount,
+		uint256 nonce
+	);
+
+	// Event to emit when a transfer is completed
+	event TransferCompleted(
+		address indexed sender,
+		address indexed recipient,
+		uint256 amount,
+		uint256 nonce
+	);
+
+	event BridgeEvent(BridgeMessage message, bytes message_bytes);
+
+	// Function to pause the bridge
+	function pauseBridge() private isRunning {
```

Comment:
TODO: for later is having this function be called by EmergencyStop() just in case of stopping the bridge in case of issues that pop up